### DISCUSSION TITLE
How can I control the execution of targets by the presence or absence of a file?

### DISCUSSION THREAD

**Discussion by kkami1115** [ropensci/targets#764] [Help] [answered]
Hi, I'm working on the analysis pipeline using targets. 

Depending on the pre-analysis status, the files that I specify using `list.files()` may not be generated.
In this situation `list.files()` creates a character with zero length, so  `tar_files()` gives me the error `cannot branch over empty target (path_to_file_files)`.
I tried to set up the function to return NULL when the length is zero, but the result was the same.

The files are not necessary for the entire analysis, so I want to run the analysis itself even without the files.
What is the most appropriate way to write a pipeline when a presence or absence of a file determines whether their child targets will build? I would like to borrow your wisdom.

The following is the reproducible example of the error.

Best,

```
library(targets)
library(tidyverse)

list_files <- function(path) {
  files = list.files(path, pattern = "hogehoge", recursive = TRUE, full.names = TRUE)
  if(length(files) > 1){
    return(files)
  }else{
    return(NULL)
  }
}
func_1 <- function(file){
  if(is.null(file)){
    return(NULL)
   }else{
    read_tsv(file = file, col_names = TRUE)
  }
}


list(
  tar_target(
    path,
    "/path/"
  ),
  tarchetypes::tar_files(
  path_to_file,
  list_files(path),
  error = "continue"
) ,
tar_target(
  tar_func_1,
  func_1(path_to_file),
  pattern = map(path_to_file),
  iteration = "list"
)
)

```

**Comment by wlandau** [comment, marked as answer]
You could remove the `tar_files()` target if the files do not exist. Sketch:

```r
# _targets.R
library(targets)
library(tarchetypes)

target_list <- list(...)
targets_files <- list(
  tarchetypes::tar_files(...),
  ...
)
files <- list.files(path, pattern = "hogehoge", recursive = TRUE, full.names = TRUE)
if (length(files) > 0L) {
  target_list <- c(target_list, targets_files)
}
target_list
```

And if the files are established before the pipeline runs, then `tar_files_input()` is nicer because  of the way `tar_files()` uses cues.

**Reply by kkami1115** [reply]
Thank you! That's a method I hadn't thought of at all. 
I took this opportunity to completely restructure of my pipeline and I think I've solved the problem.
Should I use `tar_files_input()` if I don't expect any updates to the files?

**Reply by wlandau** [reply]
> Should I use tar_files_input() if I don't expect any updates to the files?

`tar_files_input()` is pretty cheap to use, so I recommend it when possible for those cases.
